redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/widget/ProtocolBridge.kt

```diff
@@ -137,6 +169,225 @@ public class ProtocolBridge<W : Any>(
   private fun node(id: Id): ProtocolNode<W> {
     return checkNotNull(nodes[id]) { "Unknown widget ID ${id.value}" }
   }
+
+  /**
+   * Implements widget reuse (view recycling).
+   *
+   * When a widget is eligible from reuse:
+   *
+   *  * It is removed from [pool].
+   *  * It is added to [nodes], alongside its descendant nodes.
+   *
+   * Returns the updated set of changes that omits any changes that were implemented with reuse.
+   */
+  private fun applyReuse(changes: List<Change>): List<Change> {
+    val idToSubtree = mutableMapOf<Id, ReuseSubtree<W>>()
+
+    // Find nodes that have Modifier.reuse
+    for ((index, change) in changes.withIndex()) {
+      if (change !is ModifierChange) continue
+
+      val reuseModifierIndex = change.elements.indexOfFirst { it.tag.value == REUSE_MODIFIER_TAG }
+      if (reuseModifierIndex == -1) continue
```

Comment:
This also short-circuits
```suggestion
      val wantsReuse = change.elements.any { it.tag.value == REUSE_MODIFIER_TAG }
      if (!wantsReuse) continue
```

Reply:
Much better. An earlier version wanted `reuseModifierIndex` but I don’t need that any more.